lib/Model/ConvertFunctionType.cpp

```diff
@@ -0,0 +1,664 @@
+  using CC = ::abi::CallingConventionTrait<V>;
+  static constexpr model::Architecture::Values A = getArchitecture(V);
+  if (!verify<A>(Function.Arguments, CC::GeneralPurposeArgumentRegisters))
+    return std::nullopt;
+  if (!verify<A>(Function.ReturnValues, CC::GeneralPurposeReturnValueRegisters))
+    return std::nullopt;
+
+  constexpr model::Register::Values PTCRR = CC::ReturnValueLocationRegister;
+  if (PTCRR != model::Register::Invalid)
+    revng_assert(model::Register::getArchitecture(PTCRR) == A);
+  for (auto &SavedRegister : CC::CalleeSavedRegisters)
+    revng_assert(model::Register::getArchitecture(SavedRegister) == A);
+
+  model::CABIFunctionType Result;
+  Result.ABI = V;
+
+  auto ArgumentList = convertArguments<CC>(Function.Arguments,
+                                           CC::GeneralPurposeArgumentRegisters,
+                                           TheBinary);
+  if (ArgumentList == std::nullopt)
+    return std::nullopt;
+  for (auto &Argument : *ArgumentList)
+    Result.Arguments.insert(Argument);
+
+  auto
+    ReturnValue = convertReturnValue<CC>(Function.ReturnValues,
+                                         CC::GeneralPurposeReturnValueRegisters,
+                                         CC::ReturnValueLocationRegister,
+                                         TheBinary);
+  if (ReturnValue == std::nullopt)
+    return std::nullopt;
+  Result.ReturnType = *ReturnValue;
+  return Result;
+}
+
+std::optional<model::CABIFunctionType>
+convertToCABIFunctionType(const model::RawFunctionType &Function,
+                          model::abi::Values ABI,
+                          model::Binary &TheBinary) {
+  revng_assert(ABI != model::abi::Invalid);
+  return skippingEnumSwitch<1>(ABI, [&]<model::abi::Values A>() {
+    return convertToCABIFunctionTypeImpl<A>(Function, TheBinary);
+  });
+}
+
+//
+// toRaw
+//
+
+using IndexType = decltype(model::Argument::Index);
+using RegisterList = llvm::SmallVector<model::Register::Values, 1>;
+struct DistributedArgument {
+  RegisterList Registers = {};
+  size_t Size = 0, SizeOnStack = 0;
+};
+using DistributedArguments = llvm::DenseMap<IndexType, DistributedArgument>;
+
+using ArgumentContainer = SortedVector<model::Argument>;
+template<typename CallConv>
+static std::optional<DistributedArguments>
+distributeArgumentsBasedOnPositions(const ArgumentContainer &Arguments) {
+  DistributedArguments Result;
+
+  for (const model::Argument &Argument : Arguments) {
+    auto &Distributed = Result[Argument.Index];
+
+    auto MaybeSize = Argument.Type.size();
+    if (MaybeSize == std::nullopt)
+      return std::nullopt;
+    Distributed.Size = *MaybeSize;
+
+    if (Argument.Type.isFloat()) {
+      if (Argument.Index < CallConv::VectorArgumentRegisters.size()) {
+        auto Register = CallConv::VectorArgumentRegisters[Argument.Index];
+        Distributed.Registers.emplace_back(Register);
+      } else {
+        Distributed.SizeOnStack = Distributed.Size;
```

Comment:
* Any reason not to have `bool OnStack` instead of `SizeOnStack`?
* No padding/alignment considerations for arguments ending up on the stack?

Reply:
1. The wording is quite a bit off, I think. You might be misunderstanding what this thing is for.
This is the way I support arguments (think, structs) that get split between the registers and the stack.
`Distributed.Registers` lists used registers, `Distributed.Size` stores the total size of the argument and `Distributed.SizeOnStack` - the size of whatever didn't fit into the registers. Technically, it could be recalculated when used, but I'm afraid that could technically interfere with some padding styles, etc, so...
2. I let the padding alignment be handled later down the line, when the stack allocation is actually decided (for now it's just a placeholder, but yeah, one play for every stack argument, be it a full one or a a piece of it.

Reply:
1. ACK. As long as there's a reason, I'm fine with it!
2. Uhm, OK.